google/cloud/pubsublite/internal/commit_subscriber.cc

```diff
@@ -0,0 +1,189 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsublite/internal/commit_subscriber.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+
+namespace google {
+namespace cloud {
+namespace pubsublite_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+using google::cloud::pubsublite::v1::Cursor;
+using google::cloud::pubsublite::v1::InitialCommitCursorRequest;
+using google::cloud::pubsublite::v1::StreamingCommitCursorRequest;
+using google::cloud::pubsublite::v1::StreamingCommitCursorResponse;
+
+CommitSubscriber::CommitSubscriber(
+    absl::FunctionRef<std::unique_ptr<ResumableStream>(
+        StreamInitializer<
+            google::cloud::pubsublite::v1::StreamingCommitCursorRequest,
+            google::cloud::pubsublite::v1::StreamingCommitCursorResponse>)>
+        resumable_stream_factory,
+    google::cloud::pubsublite::v1::InitialCommitCursorRequest
+        initial_commit_request)
+    : initial_commit_request_{std::move(initial_commit_request)},
+      resumable_stream_{resumable_stream_factory(std::bind(
+          &CommitSubscriber::Initializer, this, std::placeholders::_1))},
+      service_composite_{resumable_stream_.get()} {}
+
+CommitSubscriber::~CommitSubscriber() {
+  future<void> shutdown = Shutdown();
+  if (!shutdown.is_ready()) {
+    GCP_LOG(WARNING) << "`Shutdown` must be called and finished before object "
+                        "goes out of scope.";
+    assert(false);
+  }
+  shutdown.get();
+}
+
+future<Status> CommitSubscriber::Start() {
+  auto start_return = service_composite_.Start();
+  Read();
+  return start_return;
+}
+
+void CommitSubscriber::Commit(Cursor cursor) {
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    if (cursor.offset() <= to_be_sent_commits_.back().offset()) {
+      return service_composite_.Abort(
+          Status(StatusCode::kFailedPrecondition,
+                 absl::StrCat(
+                     "offset is less than or equal to previous sent offsets")));
+    }
+    to_be_sent_commits_.push_back(std::move(cursor));
+    if (sending_commits_) return;
+    sending_commits_ = true;
+  }
+  SendCommits();
+}
+
+future<void> CommitSubscriber::Shutdown() {
+  return service_composite_.Shutdown();
+}
+
+void CommitSubscriber::SendCommits() {
+  StreamingCommitCursorRequest req;
+  bool service_ok = service_composite_.status().ok();
+  AsyncRoot root;
+  std::lock_guard<std::mutex> g{mu_};
+  if (to_be_sent_commits_.empty() || !service_ok) {
+    sending_commits_ = false;
+    return;
+  }
+  *req.mutable_commit()->mutable_cursor() = to_be_sent_commits_.front();
+  outstanding_commits_.push_back(std::move(to_be_sent_commits_.front()));
+  to_be_sent_commits_.pop_front();
+
+  root.get_future()
+      .then(ChainFuture(resumable_stream_->Write(std::move(req))))
+      .then([this](future<bool> res) {
+        if (!res.get()) {
+          {
+            std::lock_guard<std::mutex> g{mu_};
```

Comment:
this is weird that this is part of SendCommits instead of part of the initializer. Can you move it there?